### PR TITLE
Implement Metadatav3

### PR DESCRIPTION
**Pull request by pawanjay176**
## Issue Addressed

Resolves https://github.com/sigp/lighthouse/issues/6233

## Proposed Changes

Add metadatav3 as defined in the spec.

### CODE REVIEW

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs [hunk 1]
```diff
@@ -100,14 +102,16 @@ impl<E: EthSpec> PeerInfo<E> {
...
-                    // as the peer count is low and it's important to identify supernodes to get a
-                    // good distribution of peers across subnets.
-                    return true;
+                Subnet::DataColumn(column) => {
+                    let Ok(custody_subnet_count) = meta_data.custody_subnet_count() else {
+                        return false;
+                    };
+                    // Supernode will be subscribed to all subnets
+                    if *custody_subnet_count == spec.data_column_sidecar_subnet_count {
+                        true
```
Comment (dapplion) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs]: Why is it necessary to handle the super node case differently? Just have a hashset with all subnets. No need to pass the chainspec then

File: beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs [hunk 2]
```diff
@@ -984,6 +998,15 @@ mod tests {
...
     }
 
+    // fn metadata_v3() -> MetaData<Spec> {
+    //     MetaData::V3(MetaDataV3 {
+    //         seq_number: 1,
+    //         attnets: EnrAttestationBitfield::<Spec>::default(),
+    //         syncnets: EnrSyncCommitteeBitfield::<Spec>::default(),
+    //         custody_subnet_count: 1,
+    //     })
+    // }
```
Comment (dapplion) [inline beacon_node/lighthouse_network/src/rpc/codec/ssz_snappy.rs]: Idle comment

File: beacon_node/lighthouse_network/src/rpc/methods.rs [hunk 3]
```diff
@@ -152,13 +164,37 @@ impl<E: EthSpec> MetaData<E> {
...
             md @ MetaData::V2(_) => md.clone(),
+            MetaData::V3(metadata) => MetaData::V2(MetaDataV2 {
+                seq_number: metadata.seq_number,
+                attnets: metadata.attnets.clone(),
+                syncnets: metadata.syncnets.clone(),
+            }),
+        }
+    }
+
+    /// Returns a V2 MetaData response from self by filling setting 0 custody.
```
Comment (dapplion) [inline beacon_node/lighthouse_network/src/rpc/methods.rs]: Wrong doc comment, and inconsistent return of variants, plus code sets to 1 not 0. Where is this function used? Should default to custody requirement

File: beacon_node/lighthouse_network/src/rpc/protocol.rs [hunk 4]
```diff
@@ -392,9 +395,19 @@ impl SupportedProtocol {
             ProtocolId::new(Self::BlocksByRootV2, Encoding::SSZSnappy),
             ProtocolId::new(Self::BlocksByRootV1, Encoding::SSZSnappy),
             ProtocolId::new(Self::PingV1, Encoding::SSZSnappy),
-            ProtocolId::new(Self::MetaDataV2, Encoding::SSZSnappy),
-            ProtocolId::new(Self::MetaDataV1, Encoding::SSZSnappy),
```
Comment (dapplion) [inline beacon_node/lighthouse_network/src/rpc/protocol.rs:396]: Comment that you need v3 to go first for priority negotiation

File: beacon_node/lighthouse_network/src/service/utils.rs [hunk 5]
```diff
@@ -276,10 +286,12 @@ pub(crate) fn save_metadata_to_disk<E: EthSpec>(
...
     let _ = std::fs::create_dir_all(dir);
-    let metadata_bytes = match metadata {
-        MetaData::V1(md) => md.as_ssz_bytes(),
-        MetaData::V2(md) => md.as_ssz_bytes(),
-    };
+    // We always store the metadata v2 to disk because
+    // custody_subnet_count parameter doesn't need to be persisted across runs.
+    // custody_subnet_count is what the user sets it for the current run.
+    // This is to prevent ugly branching logic when reading the metadata from disk.
+    // TODO(pawan): check if this is unnecessaarily convoluted.
```
Comment (dapplion) [inline beacon_node/lighthouse_network/src/service/utils.rs]: Sounds reasonable to me

**Review by dapplion — COMMENTED** [review]
Looks good overall! Some minor comments

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs [hunk 6]
```diff
@@ -364,8 +368,32 @@ impl<E: EthSpec> PeerInfo<E> {
...
+    ) {
+        // If we don't have a node id, we cannot compute the custody duties anyway
+        let Some(node_id) = node_id_opt else {
+            self.meta_data = Some(meta_data);
+            return;
+        };
+
+        // Already set by enr if custody_subnets is non empty
+        if self.custody_subnets.is_empty() {
+            if let Ok(custody_subnet_count) = meta_data.custody_subnet_count() {
```
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs:375]: we should verify if the `meta_data.custody_subnet_count` provided is valid (`<= spec.data_column_sidecar_subnet_count`).

Comment (jimmygchen) [reply beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs:375]: I've added validation in this PR
https://github.com/sigp/lighthouse/pull/6308

File: beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs [hunk 7]
```diff
@@ -89,6 +90,7 @@ impl<E: EthSpec> PeerInfo<E> {
     }
 
     /// Returns if the peer is subscribed to a given `Subnet` from the metadata attnets/syncnets field.
+    /// Also returns true if
```
Comment (jimmygchen) [inline beacon_node/lighthouse_network/src/peer_manager/peerdb/peer_info.rs]:
```suggestion
    /// Also returns true if the peer is assigned to custody a given data column `Subnet` computed from the metadata `custody_column_count` field or ENR `csc` field.
```

**Review by jimmygchen — APPROVED** [review]
Nice, looks good to me and tests are passing! 🎉 
I'll just commit the minor doc update and merge this.

**Comment by jimmygchen** [comment]
@mergify queue

**Pull request merged**